internal/jimm/controller.go

```diff
@@ -505,11 +514,28 @@ func (j *JIMM) ImportModel(ctx context.Context, user *openfga.User, controllerNa
 	model.CloudRegionID = cr.ID
 	model.CloudRegion = cr
 
-	err = j.Database.AddModel(ctx, &model)
-	if err != nil {
-		if errors.ErrorCode(err) == errors.CodeAlreadyExists {
-			return errors.E(op, err, "model already exists")
+	err = j.Database.Transaction(func(d *db.Database) error {
+		err := j.Database.AddModel(ctx, &model)
+		if err != nil {
+			if errors.ErrorCode(err) == errors.CodeAlreadyExists {
+				return errors.E(op, err, "model already exists")
+			}
+			return errors.E(op, err)
+		}
+		for _, offer := range offersToAdd {
+			var dbOffer dbmodel.ApplicationOffer
+			dbOffer.FromJujuApplicationOfferAdminDetailsV5(offer)
+			dbOffer.ModelID = model.ID
+			if err := j.Database.AddApplicationOffer(ctx, &dbOffer); err != nil {
+				if errors.ErrorCode(err) == errors.CodeAlreadyExists {
+					return errors.E(op, err, fmt.Sprintf("offer with URL %s already exists", offer.OfferURL))
+				}
+				return errors.E(op, err)
+			}
 		}
+		return nil
+	})
+	if err != nil {
```

Comment:
question: we run handleModelDeltas on L516, which will connect to the controller, create an all watcher and process a single set of deltas.. is that the intent?

Reply:
I can't say why it's like that, but that does seem intentional and has been that way for some time, the tests also verify the delta functionality.
The tests show that the deltas are used to update some properties of the model like the status fields, etc.